### PR TITLE
Require approval before attempting merges

### PR DESCRIPTION
**Pull request by ZainRizvi**
Get pytorchbot to check the approval status of a PR before it forwards a merge request to the trymerge workflow.

This allows devs to know sooner if their merge request is invalid (trymerge already checks the same things, but can take an extra few minutes before offering that feedback)

### CODE REVIEW

File: torchci/lib/bot/pytorchBotHandler.ts [hunk 1]
```diff
@@ -157,46 +222,55 @@ The explanation needs to be clear on why this is needed. Here are some good exam
...
+    } else {
+      // Ensure the PR has been signed off on
+      let isApprovedPR = await this.doesHaveApprovedStatus()
+      if (!isApprovedPR) {
+        rejection_reason = "This PR needs to be approved by an authorized maintainer before merge."
+      }
+    }
+
+    if (rejection_reason) {
+      await this.logger.log("merge-error", extra_data);
```
Comment (ZainRizvi) [inline torchci/lib/bot/pytorchBotHandler.ts:234]: note: from this point down in this file it's just a minor refactor of inverting the old `if (!rejection_reason)` block

File: torchci/test/mergeCommands.test.ts [hunk 2]
```diff
@@ -796,44 +805,9 @@ describe("merge-bot", () => {
         );
         return true;
       })
-      .reply(200, {});
-    await probot.receive(event);
-
-    handleScope(scope);
-  });
-
-  test("merge on green using CLI", async () => {
```
Comment (ZainRizvi) [inline torchci/test/mergeCommands.test.ts:805]: There were two copies of this test in the file. Removing the dupe

File: torchci/test/mergeCommands.test.ts [hunk 3]
```diff
@@ -106,7 +106,10 @@ describe("merge-bot", () => {
         );
         return true;
       })
-      .reply(200, {});
+      .reply(200, {})
```
Comment (ZainRizvi) [inline torchci/test/mergeCommands.test.ts:109]: Aside: Mocking new paths is starting to require too much boilerplate in these tests. We could use a new abstraction here.

File: torchci/test/fixtures/pull_request_reviews.json [hunk 4]
```diff
@@ -0,0 +1,119 @@
+[
```
Comment (seemethere) [inline torchci/test/fixtures/pull_request_reviews.json:1]: Might be good to add this as a generated file to .gitattributes so these don't show up in PR review

**Review by seemethere — APPROVED** [review]

File: torchci/test/fixtures/pull_request_reviews.json (hunk 4 above)
Comment (ZainRizvi) [reply torchci/test/fixtures/pull_request_reviews.json:1]: Can these be autogenerated? I copy/pasta-ed this file from a github api

**Comment by malfet** [comment]
IMO we should really stop duplicating mergebot functionality in probot. If somebody wants to rewrite it in typescript, please go ahead and do it, but otherwise creating more and more rules on the probot side that mimic what merge bot is doing making it harder to maintain and require fixing the same bugs in two places. Please either move the logic completely to mergebot (though I'm not sure how it could be done without compromising the security) or keep the logic in mergebot

**Pull request merged**